In [ ]:
import pandas as pd # Load the Python package panda for data manipulation
import numpy as np  # Load the Python package numpy for some math|
import matplotlib.pyplot as plt # this package is used make plots
import statsmodels.api as sm

##Exercise 1: Operations Excellence

##Question 1: Write the optimization models for each plant, i.e., make sure to detail the decision variables, the objective function and the constraints.
This Markdown text was re-written by using Lecture 3 - Online Advertising problem and AI

## Modeling the Optimization for BIC and CIB Sites

### Decision Variables
Let $i=\{1,2,3,4\}$ be the index for the production units such that:
- $i=1$ is Standard Units at BIC
- $i=2$ is Deluxe Units at BIC
- $i=3$ is Standard Units at CIB
- $i=4$ is Deluxe Units at CIB

Each $x_i$ represents the number of units produced in each category at each site. For instance, $x_1$ is the number of standard units produced at the BIC site.

### Objectives

#### BIC Site Objective
The objective for the BIC site is to maximize the profit contribution from producing standard and deluxe products:
\begin{equation}
Maximize \quad Z_{BIC} = 10x_1 + 15x_2
\end{equation}
- Here, $10$ and $15$ are the profit contributions per unit for Standard and Deluxe products at the BIC site, respectively.

#### CIB Site Objective
The objective for the CIB site is similarly structured to maximize its profit contribution:
\begin{equation}
Maximize \quad Z_{CIB} = 10x_3 + 15x_4
\end{equation}
- Similarly, $10$ and $15$ represent the profit per unit for Standard and Deluxe products at the CIB site, respectively.

### Constraints

#### BIC Site Constraints
\begin{align}
4x_1 + 2x_2 &\leq 80 \quad \text{(Cooking Capacity)}\\
2x_1 + 5x_2 &\leq 60 \quad \text{(Filtering Capacity)}\\
4(x_1 + x_2) &\leq 75 \quad \text{(Raw Material Usage)}
\end{align}

#### CIB Site Constraints
\begin{align}
5x_3 + 3x_4 &\leq 60 \quad \text{(Cooking Capacity)}\\
5x_3 + 6x_4 &\leq 75 \quad \text{(Filtering Capacity)}\\
4(x_3 + x_4) &\leq 45 \quad \text{(Raw Material Usage)}
\end{align}

### Non-negativity Constraints
To ensure that production quantities are feasible:
\begin{align}
x_1, x_2, x_3, x_4 \geq 0
\end{align}


##Question 2: Provide the optimal solutions and profit levels for each factory. Give the values of the decision variables and the optimal objective function value.
  
Re-written using Solution with CVXPY from Lecture 3- lego New Problem and AI

In [ ]:
pip install cvxpy


In [24]:
import cvxpy as cvx

# Define decision variables for BIC site
x1 = cvx.Variable(integer=True)  # Standard units at BIC
x2 = cvx.Variable(integer=True)  # Deluxe units at BIC

# Define decision variables for CIB site
y1 = cvx.Variable(integer=True)  # Standard units at CIB
y2 = cvx.Variable(integer=True)  # Deluxe units at CIB

# Define the objective functions
objective_bic = cvx.Maximize(10 * x1 + 15 * x2)  # Maximize profit at BIC
objective_cib = cvx.Maximize(10 * y1 + 15 * y2)  # Maximize profit at CIB

# Constraints for BIC site
constraints_bic = [
    4 * x1 + 2 * x2 <= 80,  # Cooking capacity at BIC
    2 * x1 + 5 * x2 <= 60,  # Filtering capacity at BIC
    4 * (x1 + x2) <= 75,   # Raw material usage at BIC
    x1 >= 0,               # Non-negativity for x1
    x2 >= 0                # Non-negativity for x2
]

# Constraints for CIB site
constraints_cib = [
    5 * y1 + 3 * y2 <= 60,  # Cooking capacity at CIB
    5 * y1 + 6 * y2 <= 75,  # Filtering capacity at CIB
    4 * (y1 + y2) <= 45,   # Raw material usage at CIB
    y1 >= 0,               # Non-negativity for y1
    y2 >= 0                # Non-negativity for y2
]

# Set up the problems
prob_bic = cvx.Problem(objective_bic, constraints_bic)
prob_cib = cvx.Problem(objective_cib, constraints_cib)

# Solve the problems
result_bic = prob_bic.solve()
result_cib = prob_cib.solve()

# Output results
print('Optimal profit for BIC: $', round(prob_bic.value, 2))
print('Optimal production at BIC: Standard Units =', round(x1.value.item()), ', Deluxe Units =', round(x2.value.item()))
print('Optimal profit for CIB: $', round(prob_cib.value, 2))
print('Optimal production at CIB: Standard Units =', round(y1.value.item()), ', Deluxe Units =', round(y2.value.item()))



Optimal profit for BIC: $ 220.0
Optimal production at BIC: Standard Units = 10 , Deluxe Units = 8
Optimal profit for CIB: $ 165.0
Optimal production at CIB: Standard Units = 0 , Deluxe Units = 11


##Question 3:Now, write the optimization model for the joint optimization of both sites, and provide the optimal solution and profit levels.
 Re-written using the above Markdown txt and code and AI

## Joint Optimization Model for BIC and CIB Sites

### Decision Variables
Let's define the decision variables as:
- **X1**: Number of Standard Units produced at the BIC site.
- **X2**: Number of Deluxe Units produced at the BIC site.
- **Y1**: Number of Standard Units produced at the CIB site.
- **Y2**: Number of Deluxe Units produced at the CIB site.

### Objective Function
The goal is to maximize the combined profits from both sites:
\begin{equation}
Maximize \quad Z = 10x_1 + 15x_2 + 10y_1 + 15y_2
\end{equation}
Here, the coefficients \(10\) and \(15\) represent the profit contributions per unit for the standard and deluxe products, respectively.

### Constraints
The model includes constraints for production capacities (cooking and filtering) at each site and a combined constraint for raw material usage across both sites.

#### BIC Site Constraints
\begin{align}
4x_1 + 2x_2 &\leq 80 \quad \text{(Cooking Capacity)}\\
2x_1 + 5x_2 &\leq 60 \quad \text{(Filtering Capacity)}
\end{align}

#### CIB Site Constraints
\begin{align}
5y_1 + 3y_2 &\leq 60 \quad \text{(Cooking Capacity)}\\
5y_1 + 6y_2 &\leq 75 \quad \text{(Filtering Capacity)}
\end{align}

#### Combined Raw Material Usage
\begin{align}
4(x_1 + x_2 + y_1 + y_2) &\leq 120 \quad \text{(Total Raw Material Usage)}
\end{align}

### Non-negativity Constraints
All decision variables must be non-negative:
\begin{equation}
x_1, x_2, y_1, y_2 \geq 0
\end{equation}

In [ ]:
import cvxpy as cvx

# Define decision variables
X1 = cvx.Variable(integer=True)  # Standard units at BIC
X2 = cvx.Variable(integer=True)  # Deluxe units at BIC
Y1 = cvx.Variable(integer=True)  # Standard units at CIB
Y2 = cvx.Variable(integer=True)  # Deluxe units at CIB

# Objective Function
objective = cvx.Maximize(10 * X1 + 15 * X2 + 10 * Y1 + 15 * Y2)

# Constraints
constraints = [
    4 * X1 + 2 * X2 <= 80,   # Cooking capacity at BIC
    2 * X1 + 5 * X2 <= 60,   # Filtering capacity at BIC
    5 * Y1 + 3 * Y2 <= 60,   # Cooking capacity at CIB
    5 * Y1 + 6 * Y2 <= 75,   # Filtering capacity at CIB
    4 * (X1 + X2 + Y1 + Y2) <= 120,  # Total raw material usage
    X1 >= 0,                 # Non-negativity for X1
    X2 >= 0,                 # Non-negativity for X2
    Y1 >= 0,                 # Non-negativity for Y1
    Y2 >= 0                  # Non-negativity for Y2
]

# Set up the problem
prob = cvx.Problem(objective, constraints)

# Solve the problem
result = prob.solve()

# Output results
print('Total optimal profit: $', round(prob.value, 2))
print('Optimal production at BIC - Standard Units:', round(X1.value.item()), ', Deluxe Units:', round(X2.value.item()))
print('Optimal production at CIB - Standard Units:', round(Y1.value.item()), ', Deluxe Units:', round(Y2.value.item()))


Total optimal profit: $ 400.0
Optimal production at BIC - Standard Units: 10 , Deluxe Units: 8
Optimal production at CIB - Standard Units: 0 , Deluxe Units: 12


##Question 4 : Please comment on the results? How do they compare to the results obtained in Question 2.

AI Generated but re-written by me

Here are the results from question 2, presented from an individual site optimization perspective. Each site was individually assessed to determine its optimal production levels. For instance, BIC was found to be most efficient producing 10 units of the standard product and 8 units of the deluxe product. On the other hand, CIB was optimized to produce 0 standard units but 11 deluxe units. While individual optimization might be underutilizing each site's capacity, a combined approach appears to allocate resources more effectively, leading to higher overall profitability.

## Question 5: In one paragraph (no more than 300 words), detail the recommendations you would provide to the CEO of the company to improve the firm’s operations?

AI Generated but re-written by me:

To enhance our operations and profitability, I recommend adopting a centralized approach to resource allocation and production planning across all manufacturing sites. The joint optimization results have shown that integrating decision-making for the BIC and CIB sites not only maximizes total profit but also optimizes the use of shared resources like raw materials. This approach can reduce inefficiencies from isolated planning and make the most of shifting resources where they are most effective.

Additionally, investing in technological upgrades at the CIB site can reduce production times and increase capacity, supporting our findings that streamlined operations lead to better performance. It's also important to regularly review and adjust production capacities and constraints based on real-time data and market demands to maintain optimal output levels.

Fostering a culture of continuous improvement and innovation will drive operational excellence and sustain our competitive edge. Implementing these strategies will require careful change management and possibly an initial capital investment, but the potential for significant long-term gains in efficiency and profitability makes the effort worthwhile.








----


##Exercise 2: Hotel L’Escargot

##Question 1: According to the regression equation given above, which variable positively affect Profitability? Which variable negatively affect Profitability? Does this intuitively make sense? Why? ***AI RE-Written***

Based on the regression equation provided:

Positive Influences on Profitability:
Square Root of Median Income (normalized)

Coefficient: +5.86
Higher median income in the area suggests that people might have more disposable income, leading to higher spending on accommodations, and thereby increasing profitability.
State Population Per Inn (normalized)

Coefficient: +1.75
A higher state population per inn indicates fewer inns relative to the population, suggesting less competition and potentially higher occupancy rates, which could positively impact profitability.
Negative Influences on Profitability:
Price (normalized)

Coefficient: -5.41
Higher normalized prices are associated with lower profitability. This could be because high prices might reduce demand or occupancy rates, thus negatively impacting profitability.
College Students in Area (normalized)

Coefficient: -3.09
A higher number of college students in the area negatively affects profitability. This might be because college students generally have lower spending power and may not be the primary customer base for higher-end inns or hotels.

Square Root of Median Income (normalized): The positive relationship makes sense as higher-income areas are likely to have guests willing to spend more on accommodations, increasing profitability.

Price (normalized): Although it may seem counterintuitive initially, this negative relationship aligns with the idea that overly high prices can deter potential guests, reducing occupancy rates and overall profitability.



##Question 2: Using this regression equation, L’Escargot created a spreadsheet model to predict profitability. L’Escargot collected data for several locations in California, which is provided in the excel spreadsheet on Canvas “L’Escargot”. Using this spreadsheet, compute the profitability for each hotel. Which one has the highest profitability? Which one has the lowest profitability?

Below is AI written


### Profitability Prediction for L’Escargot Hotels

Using the given regression equation:

\[ \text{Profitability} = 39.05 - (5.41 \times \text{Price (normalized)}) + (5.86 \times \text{Square Root of Median Income (normalized)}) - (3.09 \times \text{College Students in Area (normalized)}) + (1.75 \times \text{State Population Per Inn (normalized)}) \]

We computed the profitability for each hotel location in California. Below are the predicted profitability values along with the hotel having the highest and lowest profitability.

#### Predicted Profitability for Each Hotel

| Hotel | Location            | Predicted Profitability |
|-------|----------------------|-------------------------|
| 1     | Eureka, California   | 35.83                   |
| 2     | Fresno, California   | 25.67                   |
| 3     | Fresno, California   | 35.24                   |
| 4     | Fresno, California   | 35.62                   |
| 5     | Fresno, California   | 40.48                   |
| ...   | ...                  | ...                     |

#### Highest and Lowest Profitability

- **Highest Profitability**: Hotel 5, Fresno, California with a predicted profitability of 40.48
- **Lowest Profitability**: Hotel 2, Fresno, California with a predicted profitability of 25.67


Below is AI written

In [ ]:
import pandas as pd

# Load the Excel file from Google Colab
from google.colab import files
uploaded = files.upload()

# Assuming the file is named 'Lescargot.xlsx'
file_name = 'Lescargot.xlsx'
lescargot_data = pd.read_excel(file_name, header=2)

# Define the regression model coefficients
intercept = 39.05
coefficients = {
    'Price (normalized)': -5.41,
    'Square Root of Median Income (normalized)': 5.86,
    'College Students in Area (normalized)': -3.09,
    'State Population Per Inn (normalized)': 1.75
}

# Calculate the predicted profitability
lescargot_data['Predicted Profitability'] = (
    intercept +
    coefficients['Price (normalized)'] * lescargot_data['Price (normalized)'] +
    coefficients['Square Root of Median Income (normalized)'] * lescargot_data['Square Root of Median Income (normalized)'] +
    coefficients['College Students in Area (normalized)'] * lescargot_data['College Students in Area (normalized)'] +
    coefficients['State Population Per Inn (normalized)'] * lescargot_data['State Population Per Inn (normalized)']
)

# Identify the hotel with the highest and lowest profitability
highest_profitability = lescargot_data.loc[lescargot_data['Predicted Profitability'].idxmax()]
lowest_profitability = lescargot_data.loc[lescargot_data['Predicted Profitability'].idxmin()]

# Display the results
highest_profitability, lowest_profitability


Saving Lescargot.xlsx to Lescargot (1).xlsx


/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


(Hotel                                                         5
 Location                                     Fresno, California
 Price                                                    325000
 Price (normalized)                                    -1.037136
 Square Root of Median Income (normalized)             -0.408199
 College Students in Area (normalized)                  0.311669
 State Population Per Inn (normalized)                 -0.474279
 Predicted Profitability                               40.475817
 Name: 4, dtype: object,
 Hotel                                                         2
 Location                                     Fresno, California
 Price                                                  10000000
 Price (normalized)                                     1.699076
 Square Root of Median Income (normalized)             -0.408199
 College Students in Area (normalized)                  0.311669
 State Population Per Inn (normalized)                 -0.474279


##Question 3: L’Escargot has a budget of $10 million to spend on hotels. Suppose we use a “greedy” approach where we select the most profitable hotels until we ran out of budget. So we would start by the most profitable, and then if we had enough budget left, we would buy the hotel we predict to be the second most profitable, and so on.


##Part A : Describe what we would do with this approach, i.e., which hotels would we purchase? ***AI re-written***


To implement the "greedy" approach, we will start by sorting the list of hotels in descending order based on their predicted profitability. Next, we will iterate through the sorted list, selecting each hotel if its price, when added to the current total cost, does not exceed the budget limit of 10 million. We will continue this process, starting with the most profitable hotel and adding its cost to the total budget spent, until we reach the 10 million budget limit. We will stop once adding another hotel would exceed this budget.

AI Generated:

### Greedy Approach for Selecting Hotels within a $10 Million Budget

Using the "greedy" approach, we select the most profitable hotels until we run out of budget. Here are the steps we follow:

1. **Sort the Hotels by Predicted Profitability**: Sort the list of hotels in descending order based on their predicted profitability.
2. **Select Hotels Until Budget is Exhausted**: Starting from the most profitable hotel, add the hotel's cost to the total budget spent until the budget of $10 million is exhausted.

#### Selected Hotels

After applying this approach, the selected hotels are as follows:

| Hotel | Location            | Price     | Predicted Profitability |
|-------|----------------------|-----------|-------------------------|
| X     | Location X           | $X        | XX.XX                   |
| Y     | Location Y           | $Y        | YY.YY                   |
| ...   | ...                  | ...       | ...                     |

The above table lists the hotels we would purchase using the "greedy" approach within the $10 million budget.


AI Generated:

In [ ]:
import pandas as pd

# Load the Excel file from Google Colab
from google.colab import files
uploaded = files.upload()

# Assuming the file is named 'Lescargot.xlsx'
file_name = 'Lescargot.xlsx'
lescargot_data = pd.read_excel(file_name, header=2)

# Define the regression model coefficients
intercept = 39.05
coefficients = {
    'Price (normalized)': -5.41,
    'Square Root of Median Income (normalized)': 5.86,
    'College Students in Area (normalized)': -3.09,
    'State Population Per Inn (normalized)': 1.75
}

# Calculate the predicted profitability
lescargot_data['Predicted Profitability'] = (
    intercept +
    coefficients['Price (normalized)'] * lescargot_data['Price (normalized)'] +
    coefficients['Square Root of Median Income (normalized)'] * lescargot_data['Square Root of Median Income (normalized)'] +
    coefficients['College Students in Area (normalized)'] * lescargot_data['College Students in Area (normalized)'] +
    coefficients['State Population Per Inn (normalized)'] * lescargot_data['State Population Per Inn (normalized)']
)

# Sort the hotels by predicted profitability in descending order
sorted_hotels = lescargot_data.sort_values(by='Predicted Profitability', ascending=False)

# Initialize variables for the greedy approach
budget = 10_000_000  # $10 million
total_cost = 0
selected_hotels = []

# Select hotels until the budget is exhausted
for index, row in sorted_hotels.iterrows():
    price = row['Price']
    if total_cost + price <= budget:
        selected_hotels.append(row)
        total_cost += price

# Create a DataFrame of selected hotels
selected_hotels_df = pd.DataFrame(selected_hotels)

# Display the selected hotels
selected_hotels_df[['Hotel', 'Location', 'Price', 'Predicted Profitability']]


Saving Lescargot.xlsx to Lescargot (2).xlsx


/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


Hotel                      Location    Price  Predicted Profitability
4       5            Fresno, California   325000                40.475817
15     16  South Lake Tahoe, California   750000                40.466618
10     11  South Lake Tahoe, California  1125000                39.892862
14     15  South Lake Tahoe, California  1475000                39.357356
7       8       Los Angeles, California  1750000                39.301497
9      10  South Lake Tahoe, California  1650000                39.089603
6       7       Los Angeles, California  1950000                38.995493

##Part B: What would our total predicted profitability be? (This is the sum of the predicted profitability of all hotels we purchase.)


AI Generated:

### Greedy Approach for Selecting Hotels within a $10 Million Budget

Using the "greedy" approach, we select the most profitable hotels until we run out of budget. Here are the steps we follow:

1. **Sort the Hotels by Predicted Profitability**: Sort the list of hotels in descending order based on their predicted profitability.
2. **Select Hotels Until Budget is Exhausted**: Starting from the most profitable hotel, add the hotel's cost to the total budget spent until the budget of $10 million is exhausted.

#### Selected Hotels

After applying this approach, the selected hotels are as follows:

| Hotel | Location                   | Price     | Predicted Profitability |
|-------|-----------------------------|-----------|-------------------------|
| 5     | Fresno, California          | 325,000   | 40.48                   |
| 16    | South Lake Tahoe, California| 750,000   | 40.47                   |
| 11    | South Lake Tahoe, California| 1,125,000 | 39.89                   |
| 15    | South Lake Tahoe, California| 1,475,000 | 39.36                   |
| 8     | Los Angeles, California     | 1,750,000 | 39.30                   |
| 10    | South Lake Tahoe, California| 1,650,000 | 39.09                   |
| 7     | Los Angeles, California     | 1,950,000 | 39.00                   |

#### Total Predicted Profitability

The total predicted profitability of the selected hotels is: **277.58**


AI Generated:

In [ ]:
import pandas as pd

# Load the Excel file from Google Colab
from google.colab import files
uploaded = files.upload()

# Assuming the file is named 'Lescargot.xlsx'
file_name = 'Lescargot.xlsx'
lescargot_data = pd.read_excel(file_name, header=2)

# Define the regression model coefficients
intercept = 39.05
coefficients = {
    'Price (normalized)': -5.41,
    'Square Root of Median Income (normalized)': 5.86,
    'College Students in Area (normalized)': -3.09,
    'State Population Per Inn (normalized)': 1.75
}

# Calculate the predicted profitability
lescargot_data['Predicted Profitability'] = (
    intercept +
    coefficients['Price (normalized)'] * lescargot_data['Price (normalized)'] +
    coefficients['Square Root of Median Income (normalized)'] * lescargot_data['Square Root of Median Income (normalized)'] +
    coefficients['College Students in Area (normalized)'] * lescargot_data['College Students in Area (normalized)'] +
    coefficients['State Population Per Inn (normalized)'] * lescargot_data['State Population Per Inn (normalized)']
)

# Sort the hotels by predicted profitability in descending order
sorted_hotels = lescargot_data.sort_values(by='Predicted Profitability', ascending=False)

# Initialize variables for the greedy approach
budget = 10_000_000  # $10 million
total_cost = 0
selected_hotels = []

# Select hotels until the budget is exhausted
for index, row in sorted_hotels.iterrows():
    price = row['Price']
    if total_cost + price <= budget:
        selected_hotels.append(row)
        total_cost += price

# Create a DataFrame of selected hotels
selected_hotels_df = pd.DataFrame(selected_hotels)

# Calculate the total predicted profitability
total_predicted_profitability = selected_hotels_df['Predicted Profitability'].sum()

# Display the selected hotels and total predicted profitability
selected_hotels_df[['Hotel', 'Location', 'Price', 'Predicted Profitability']], total_predicted_profitability


Saving Lescargot.xlsx to Lescargot (4).xlsx


/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


(    Hotel                      Location    Price  Predicted Profitability
 4       5            Fresno, California   325000                40.475817
 15     16  South Lake Tahoe, California   750000                40.466618
 10     11  South Lake Tahoe, California  1125000                39.892862
 14     15  South Lake Tahoe, California  1475000                39.357356
 7       8       Los Angeles, California  1750000                39.301497
 9      10  South Lake Tahoe, California  1650000                39.089603
 6       7       Los Angeles, California  1950000                38.995493,
 277.57924735275594)

#Part C: If we are trying to maximize our total predicted profitability, is this a good approach? How about if we were trying to maximize the average predicted profitability of the hotels we select? How about if we had a budget of 20 million instead of 10 million?

AI Generated:

### Evaluation of the Greedy Approach for Hotel Selection

#### Maximizing Total Predicted Profitability

Using the greedy approach, we select the most profitable hotels until the budget is exhausted. This heuristic is generally effective in maximizing total predicted profitability, but it may not always yield the optimal solution.

#### Maximizing Average Predicted Profitability

If our goal is to maximize the average predicted profitability, the greedy approach might not be ideal. Instead, we should consider both profitability and cost, ensuring we select a mix of hotels that offers the highest average return.

#### Budget Scenarios

We evaluated two budget scenarios: $10 million and $20 million.

##### Budget: $10 million

| Hotel | Location                   | Price     | Predicted Profitability |
|-------|-----------------------------|-----------|-------------------------|
| 5     | Fresno, California          | 325,000   | 40.48                   |
| 16    | South Lake Tahoe, California| 750,000   | 40.47                   |
| 11    | South Lake Tahoe, California| 1,125,000 | 39.89                   |
| 15    | South Lake Tahoe, California| 1,475,000 | 39.36                   |
| 8     | Los Angeles, California     | 1,750,000 | 39.30                   |
| 10    | South Lake Tahoe, California| 1,650,000 | 39.09                   |
| 7     | Los Angeles, California     | 1,950,000 | 39.00                   |

**Total Predicted Profitability**: 277.58  
**Average Predicted Profitability**: 39.65

##### Budget: $20 million

| Hotel | Location                   | Price     | Predicted Profitability |
|-------|-----------------------------|-----------|-------------------------|
| (same format as above for the selected hotels) |

**Total Predicted Profitability**: (result from code)  
**Average Predicted Profitability**: (result from code)

These evaluations provide insights into how different budget constraints impact the selection of hotels and their total and average predicted profitability.


AI Generated:

In [ ]:
import pandas as pd

# Load the Excel file from Google Colab
from google.colab import files
uploaded = files.upload()

# Assuming the file is named 'Lescargot.xlsx'
file_name = 'Lescargot.xlsx'
lescargot_data = pd.read_excel(file_name, header=2)

# Define the regression model coefficients
intercept = 39.05
coefficients = {
    'Price (normalized)': -5.41,
    'Square Root of Median Income (normalized)': 5.86,
    'College Students in Area (normalized)': -3.09,
    'State Population Per Inn (normalized)': 1.75
}

# Calculate the predicted profitability
lescargot_data['Predicted Profitability'] = (
    intercept +
    coefficients['Price (normalized)'] * lescargot_data['Price (normalized)'] +
    coefficients['Square Root of Median Income (normalized)'] * lescargot_data['Square Root of Median Income (normalized)'] +
    coefficients['College Students in Area (normalized)'] * lescargot_data['College Students in Area (normalized)'] +
    coefficients['State Population Per Inn (normalized)'] * lescargot_data['State Population Per Inn (normalized)']
)

# Function to select hotels based on budget
def select_hotels(budget):
    # Sort the hotels by predicted profitability in descending order
    sorted_hotels = lescargot_data.sort_values(by='Predicted Profitability', ascending=False)

    # Initialize variables for the greedy approach
    total_cost = 0
    selected_hotels = []

    # Select hotels until the budget is exhausted
    for index, row in sorted_hotels.iterrows():
        price = row['Price']
        if total_cost + price <= budget:
            selected_hotels.append(row)
            total_cost += price

    # Create a DataFrame of selected hotels
    selected_hotels_df = pd.DataFrame(selected_hotels)

    # Calculate the total and average predicted profitability
    total_predicted_profitability = selected_hotels_df['Predicted Profitability'].sum()
    average_predicted_profitability = selected_hotels_df['Predicted Profitability'].mean()

    return selected_hotels_df, total_predicted_profitability, average_predicted_profitability

# Evaluate for a budget of $10 million
selected_hotels_10m, total_profit_10m, average_profit_10m = select_hotels(10_000_000)
# Evaluate for a budget of $20 million
selected_hotels_20m, total_profit_20m, average_profit_20m = select_hotels(20_000_000)

# Display the results
(selected_hotels_10m[['Hotel', 'Location', 'Price', 'Predicted Profitability']], total_profit_10m, average_profit_10m),
(selected_hotels_20m[['Hotel', 'Location', 'Price', 'Predicted Profitability']], total_profit_20m, average_profit_20m)


Saving Lescargot.xlsx to Lescargot (5).xlsx


/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


(    Hotel                      Location    Price  Predicted Profitability
 4       5            Fresno, California   325000                40.475817
 15     16  South Lake Tahoe, California   750000                40.466618
 10     11  South Lake Tahoe, California  1125000                39.892862
 14     15  South Lake Tahoe, California  1475000                39.357356
 7       8       Los Angeles, California  1750000                39.301497
 9      10  South Lake Tahoe, California  1650000                39.089603
 6       7       Los Angeles, California  1950000                38.995493
 12     13  South Lake Tahoe, California  1975000                38.592348
 11     12  South Lake Tahoe, California  2500000                37.789090
 13     14  South Lake Tahoe, California  3750000                35.876569,
 389.83725395852593,
 38.983725395852595)

##Question 4. Now, build an optimization model to select hotels given the $10 million budget.


##Part A:  Write out the optimization problem. Make sure to detail the decision variables, the objective function and the constraints

AI Re-written:

### Optimization Model for Selecting Hotels within a $10 Million Budget

We formulated and solved an optimization problem to select hotels given a $10 million budget.

#### Decision Variables
- \( x_i \): Binary variable representing whether hotel \( i \) is selected (1) or not (0).

#### Objective Function
- Maximize the total predicted profitability:
\[ \text{Maximize} \quad \sum_{i=1}^{n} p_i \cdot x_i \]

#### Constraints
1. **Budget Constraint**: The total cost of the selected hotels must not exceed $10 million:
\[ \sum_{i=1}^{n} c_i \cdot x_i \leq 10{,}000{,}000 \]

2. **Binary Decision Variables**: Each decision variable \( x_i \) must be binary:
\[ x_i \in \{0, 1\} \]

#### Results

After solving the optimization problem, the selected hotels are as follows:

| Hotel | Location                   | Price     | Predicted Profitability |
|-------|-----------------------------|-----------|-------------------------|
| ...   | ...                         | ...       | ...                     |

**Total Predicted Profitability**: (result from code)  
**Average Predicted Profitability**: (result from code)


AI Generated:

##Part B: What is the optimal solution? Give the values of the decision variables and the optimal
objective function value

In [ ]:
# Install PuLP
!pip install pulp

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.7/17.7 MB 40.2 MB/s eta 0:00:00


In [ ]:
import pandas as pd
from pulp import LpMaximize, LpProblem, LpVariable, lpSum

# Load the Excel file from Google Colab
from google.colab import files
uploaded = files.upload()

# Assuming the file is named 'Lescargot.xlsx'
file_name = 'Lescargot.xlsx'
lescargot_data = pd.read_excel(file_name, header=2)

# Define the regression model coefficients
intercept = 39.05
coefficients = {
    'Price (normalized)': -5.41,
    'Square Root of Median Income (normalized)': 5.86,
    'College Students in Area (normalized)': -3.09,
    'State Population Per Inn (normalized)': 1.75
}

# Calculate the predicted profitability
lescargot_data['Predicted Profitability'] = (
    intercept +
    coefficients['Price (normalized)'] * lescargot_data['Price (normalized)'] +
    coefficients['Square Root of Median Income (normalized)'] * lescargot_data['Square Root of Median Income (normalized)'] +
    coefficients['College Students in Area (normalized)'] * lescargot_data['College Students in Area (normalized)'] +
    coefficients['State Population Per Inn (normalized)'] * lescargot_data['State Population Per Inn (normalized)']
)

# Initialize the optimization problem
model = LpProblem(name="hotel-selection", sense=LpMaximize)

# Define decision variables
x = {i: LpVariable(name=f"x_{i}", cat="Binary") for i in lescargot_data.index}

# Define the objective function
model += lpSum(lescargot_data.loc[i, "Predicted Profitability"] * x[i] for i in lescargot_data.index), "Total Predicted Profitability"

# Add the budget constraint
model += lpSum(lescargot_data.loc[i, "Price"] * x[i] for i in lescargot_data.index) <= 10_000_000, "Budget Constraint"

# Solve the optimization problem
model.solve()

# Extract the results
selected_hotels = [i for i in lescargot_data.index if x[i].value() == 1]
selected_hotels_df = lescargot_data.loc[selected_hotels]

# Calculate the total and average predicted profitability
total_predicted_profitability = selected_hotels_df['Predicted Profitability'].sum()
average_predicted_profitability = selected_hotels_df['Predicted Profitability'].mean()

# Display the results
selected_hotels_df[['Hotel', 'Location', 'Price', 'Predicted Profitability']], total_predicted_profitability, average_predicted_profitability


Saving Lescargot.xlsx to Lescargot (6).xlsx


/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


(    Hotel                      Location    Price  Predicted Profitability
 4       5            Fresno, California   325000                40.475817
 6       7       Los Angeles, California  1950000                38.995493
 7       8       Los Angeles, California  1750000                39.301497
 9      10  South Lake Tahoe, California  1650000                39.089603
 10     11  South Lake Tahoe, California  1125000                39.892862
 14     15  South Lake Tahoe, California  1475000                39.357356
 15     16  South Lake Tahoe, California   750000                40.466618,
 277.579247352756,
 39.65417819325086)

##Part C:  Does the optimal solution make sense intuitively? How does it compared to the greedy solution?
***AI Re-written***

Yes, the optimal solution makes intuitive sense.

The optimal solution selects a combination of hotels that maximizes the total predicted profitability while staying within the $10 million budget. This approach systematically evaluates the profitability and costs of all possible combinations, ensuring that the chosen set yields the highest total profit.

Compared to the greedy solution, which selects the most profitable hotels one by one until the budget is exhausted, the optimization model finds a more balanced selection. This could include slightly less profitable but lower-cost hotels that collectively contribute to higher total profitability.

Comparison to the Greedy Solution:
Total Predicted Profitability: The optimal solution typically achieves higher total predicted profitability than the greedy solution.
Hotel Selection: The optimal solution might include hotels with slightly lower individual profitability but a better fit within the budget to maximize total profitability.
Thus, the optimal solution ensures the best possible use of the budget for maximizing profitability, aligning well with the constraints and objectives.

##Question 5. L’Escargot thinks that buying too many hotels in one city is probably not a good idea and would prefer to diversify across as many cities as possible. Add constraint(s) to your model to limit the number of hotels purchased in any city to at most 2.


##Part A : What are the constraints that you need to add to the model? Intuitively, do you expect the new optimal objective function value to be larger, smaller or the same as before? *** AI generated & ReWritten***

New Constraints to Add to the Model
To ensure that L’Escargot buys no more than two hotels in any city, we need to add constraints that limit the number of selected hotels per city to at most two.

Constraints:
For each city
𝑗
j, let
𝐶
𝑗
C
j
​
  be the set of indices of hotels located in city
𝑗
j. We need to add the following constraint for each city:

∑
𝑖
∈
𝐶
𝑗
𝑥
𝑖
≤
2
∑
i∈C
j
​

​
 x
i
​
 ≤2

Expected Change in the Optimal Objective Function Value
Intuitively, we can expect the new optimal objective function value to be smaller than before. This is because the added constraints reduce the feasible solution space, limiting our flexibility to select the most profitable hotels. By imposing a maximum of two hotels per city, the model may need to select less profitable hotels from different cities to meet the diversification requirement, potentially lowering the total predicted profitability.








##Part B: Write the new optimization model.

AI Re-written:

Optimization Model for Selecting Hotels within a $10 Million Budget (With Diversification Constraint)
We formulated and solved an optimization problem to select hotels given a 10 million budget with an additional constraint to limit the number of hotels per city to at most 2.

Decision Variables
𝑥
𝑖
x
i
​
 : Binary variable representing whether hotel
𝑖
i is selected (1) or not (0).

Objective Function
Maximize the total predicted profitability:
Maximize
∑
𝑖
=
1
𝑛
𝑝
𝑖
⋅
𝑥
𝑖
Maximize∑
i=1
n
​
 p
i
​
 ⋅x
i
​


Constraints
Budget Constraint: The total cost of the selected hotels must not exceed 10 million:
∑
𝑖
=
1
𝑛
𝑐
𝑖
⋅
𝑥
𝑖
≤
10
,
000
,
000
∑
i=1
n
​
 c
i
​
 ⋅x
i
​
 ≤10,000,000

Diversification Constraint: Limit the number of hotels per city to at most 2:
∑
𝑖
∈
𝐶
𝑗
𝑥
𝑖
≤
2
∀

city

𝑗
∑
i∈C
j
​

​
 x
i
​
 ≤2∀cityj

Binary Decision Variables: Each decision variable
𝑥
𝑖
x
i
​
  must be binary:
𝑥
𝑖
∈
{
0
,
1
}
x
i
​
 ∈{0,1}

#### Results

After solving the optimization problem, the selected hotels are as follows:

| Hotel | Location                   | Price     | Predicted Profitability |
|-------|-----------------------------|-----------|-------------------------|
| ...   | ...                         | ...       | ...                     |

**Total Predicted Profitability**: (result from code)  
**Average Predicted Profitability**: (result from code)

The values of the decision variables \( x_i \) for each selected hotel are 1, indicating that they are included in the optimal solution.
``` &#8203;``【oaicite:0】``&#8203;

##Part C: Solve the new model. Give the values of the decision variables and the optimal objective function value. How does this compare to the previous solution?

AI Generated:

In [12]:
import pandas as pd
from pulp import LpMaximize, LpProblem, LpVariable, lpSum

# Load the Excel file from Google Colab
from google.colab import files
uploaded = files.upload()

# Assuming the file is named 'Lescargot.xlsx'
file_name = 'Lescargot.xlsx'
lescargot_data = pd.read_excel(file_name, header=2)

# Define the regression model coefficients
intercept = 39.05
coefficients = {
    'Price (normalized)': -5.41,
    'Square Root of Median Income (normalized)': 5.86,
    'College Students in Area (normalized)': -3.09,
    'State Population Per Inn (normalized)': 1.75
}

# Calculate the predicted profitability
lescargot_data['Predicted Profitability'] = (
    intercept +
    coefficients['Price (normalized)'] * lescargot_data['Price (normalized)'] +
    coefficients['Square Root of Median Income (normalized)'] * lescargot_data['Square Root of Median Income (normalized)'] +
    coefficients['College Students in Area (normalized)'] * lescargot_data['College Students in Area (normalized)'] +
    coefficients['State Population Per Inn (normalized)'] * lescargot_data['State Population Per Inn (normalized)']
)

# Extract city information
lescargot_data['City'] = lescargot_data['Location'].apply(lambda x: x.split(',')[0])

# Initialize the optimization problem
model = LpProblem(name="hotel-selection", sense=LpMaximize)

# Define decision variables
x = {i: LpVariable(name=f"x_{i}", cat="Binary") for i in lescargot_data.index}

# Define the objective function
model += lpSum(lescargot_data.loc[i, "Predicted Profitability"] * x[i] for i in lescargot_data.index), "Total Predicted Profitability"

# Add the budget constraint
model += lpSum(lescargot_data.loc[i, "Price"] * x[i] for i in lescargot_data.index) <= 10_000_000, "Budget Constraint"

# Add constraints to limit the number of hotels per city to at most 2
cities = lescargot_data['City'].unique()
for city in cities:
    city_indices = lescargot_data[lescargot_data['City'] == city].index
    model += lpSum(x[i] for i in city_indices) <= 2, f"Max_2_hotels_in_{city}"

# Solve the optimization problem
model.solve()

# Extract the results
selected_hotels = [i for i in lescargot_data.index if x[i].value() == 1]
selected_hotels_df = lescargot_data.loc[selected_hotels]

# Calculate the total and average predicted profitability
total_predicted_profitability = selected_hotels_df['Predicted Profitability'].sum()
average_predicted_profitability = selected_hotels_df['Predicted Profitability'].mean()

# Display the selected hotels and total predicted profitability
selected_hotels_df[['Hotel', 'Location', 'Price', 'Predicted Profitability']], total_predicted_profitability, average_predicted_profitability


Saving Lescargot.xlsx to Lescargot (9).xlsx


/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


(    Hotel                      Location    Price  Predicted Profitability
 0       1            Eureka, California  2925000                35.834516
 4       5            Fresno, California   325000                40.475817
 6       7       Los Angeles, California  1950000                38.995493
 7       8       Los Angeles, California  1750000                39.301497
 10     11  South Lake Tahoe, California  1125000                39.892862
 15     16  South Lake Tahoe, California   750000                40.466618,
 234.96680342707901,
 39.16113390451317)

##Question 6 : In one paragraph (no more than 300 words), describe how you would present your results to L’Escargot. Do you have any recommendations for them to improve the regression model? How about to improve the optimization model? ***AI ReWRITEN***

To present the results to L’Escargot, I would start by explaining the optimization model we developed to maximize the total predicted profitability of hotel selections within a $10 million budget. I would highlight the key decision variables, the objective function, and the constraints, including the diversification constraint that limits the selection to a maximum of two hotels per city.

Next, I would present the list of selected hotels, their locations, prices, and predicted profitability, emphasizing both the total and average predicted profitability achieved. Additionally, I would compare these results to the greedy approach, showing how the optimization model provides a more balanced and potentially more profitable selection.

To improve the regression model, I recommend incorporating additional variables that might influence profitability, such as local tourism trends, seasonality effects, and competitive analysis. For the optimization model, I suggest exploring more advanced techniques like mixed-integer programming or stochastic optimization to account for uncertainty and variability in hotel profitability and costs. Additionally, implementing sensitivity analysis can provide insights into how changes in budget or constraints impact the optimal solution, enabling more informed decision-making.

----


###Exercise 3: Matching for a Dating Platform

You run an online dating app that specializes in matching people based on compatibility
scores. When signing up, each person writes down their score for 5 activities: playing sports,
going to the theater, attending religious services, enjoying the outdoors and eating out. The
scores vary from -2 (strong dislike) to +2 (strong like). The activity scores chosen by 16
people in a particular location are available here (also on Canvas):
https://raw.githubusercontent.com/ormarketing/OD/master/OD.csv
When you create a match, the value of the match is equal to the sum of the products of the 5
different activity scores of the two individuals. For example, if you match Laura and Ralph,
the match score would be:
Score(Laura & Ralph) = $0\times0+2\times2+(1)\times2+0\times2+2\times(-1) = 0+4-
2+0-2 = 0$


##Question 1: Assume all of the individuals in the data set would like to be matched to people of the opposite gender. What matches would you choose to maximize the aggregate match score?

In [22]:
import pandas as pd
import numpy as np
from scipy.optimize import linear_sum_assignment

# Load the data
url = "https://raw.githubusercontent.com/ormarketing/OD/master/OD.csv"
data = pd.read_csv(url)

# Separate individuals by gender
males = data[data['Gender'] == 'Male']
females = data[data['Gender'] == 'Female']

# Reset the index for easier handling later
males.reset_index(drop=True, inplace=True)
females.reset_index(drop=True, inplace=True)

# Initialize a matrix to store the match scores
match_scores = np.zeros((len(males), len(females)))

# Calculate match scores
for i in range(len(males)):
    for j in range(len(females)):
        match_scores[i, j] = np.dot(males.iloc[i, 2:], females.iloc[j, 2:])

# Display the match scores matrix
print("Match Scores Matrix:\n", match_scores)

# Use the Hungarian algorithm to find the optimal pairs
row_ind, col_ind = linear_sum_assignment(-match_scores)

# Display the optimal pairs and their match scores
total_score = 0
pairs = []
for i, j in zip(row_ind, col_ind):
    score = match_scores[i, j]
    total_score += score
    pairs.append((males.iloc[i]['Unnamed: 0'], females.iloc[j]['Unnamed: 0'], score))

print("Optimal Pairs and Scores:")
for pair in pairs:
    print(pair)

print("\nTotal Aggregate Match Score:", total_score)


Match Scores Matrix:
 [[ -6.  -5.   6.  -5.   0.  -1.   5.   0.]
 [  1.   2.   5.   1.  -2.  -3.  -1.   5.]
 [  3.   3.  -2.   4.  -3.   5.   4.  -3.]
 [  2.   2.   4.  -4.   4. -12.   0.   4.]
 [  1.  -1.   8.   2.  -1.  -4.  -1.  -1.]
 [  6.   6.   4.   0.   2. -12.  -2.   6.]
 [  8.   5.  -2.   7.   0.  -2.  -6.  -2.]
 [  0.  -1.   0.   5.  -4.   8.   0.  -4.]]
Optimal Pairs and Scores:
('James', 'Mary', 5.0)
('Robert', 'Sophia', 5.0)
('Peter', 'Olivia', 4.0)
('Eric', 'Martha', 4.0)
('Daniel', 'Emma', 8.0)
('Adam', 'Jennifer', 6.0)
('Carl', 'Laura', 8.0)
('Ralph', 'Isabella', 8.0)

Total Aggregate Match Score: 48.0


##Question 2: How would the results change if each person is presented with two possible matches?

AI Generated:

In [23]:
import pandas as pd
import numpy as np
from collections import defaultdict

# Load the data
url = "https://raw.githubusercontent.com/ormarketing/OD/master/OD.csv"
data = pd.read_csv(url)

# Separate individuals by gender
males = data[data['Gender'] == 'Male']
females = data[data['Gender'] == 'Female']

# Reset the index for easier handling later
males.reset_index(drop=True, inplace=True)
females.reset_index(drop=True, inplace=True)

# Initialize a matrix to store the match scores
match_scores = np.zeros((len(males), len(females)))

# Calculate match scores
for i in range(len(males)):
    for j in range(len(females)):
        match_scores[i, j] = np.dot(males.iloc[i, 2:], females.iloc[j, 2:])

# Display the match scores matrix
print("Match Scores Matrix:\n", match_scores)

# Create a dictionary to store the matches
matches = defaultdict(list)

# Find the top two matches for each male
for i in range(len(males)):
    male_scores = match_scores[i, :]
    top_two_females = np.argsort(male_scores)[-2:]  # Get indices of the top two scores
    for idx in top_two_females:
        matches[males.iloc[i]['Unnamed: 0']].append((females.iloc[idx]['Unnamed: 0'], male_scores[idx]))

# Find the top two matches for each female
for j in range(len(females)):
    female_scores = match_scores[:, j]
    top_two_males = np.argsort(female_scores)[-2:]  # Get indices of the top two scores
    for idx in top_two_males:
        matches[females.iloc[j]['Unnamed: 0']].append((males.iloc[idx]['Unnamed: 0'], female_scores[idx]))

# Display the matches
for person, match_list in matches.items():
    match_list = sorted(match_list, key=lambda x: -x[1])  # Sort matches by score descending
    print(f"{person}'s top matches:")
    for match in match_list[:2]:
        print(f"  - {match[0]} with a score of {match[1]}")


Match Scores Matrix:
 [[ -6.  -5.   6.  -5.   0.  -1.   5.   0.]
 [  1.   2.   5.   1.  -2.  -3.  -1.   5.]
 [  3.   3.  -2.   4.  -3.   5.   4.  -3.]
 [  2.   2.   4.  -4.   4. -12.   0.   4.]
 [  1.  -1.   8.   2.  -1.  -4.  -1.  -1.]
 [  6.   6.   4.   0.   2. -12.  -2.   6.]
 [  8.   5.  -2.   7.   0.  -2.  -6.  -2.]
 [  0.  -1.   0.   5.  -4.   8.   0.  -4.]]
James's top matches:
  - Emma with a score of 6.0
  - Mary with a score of 5.0
Robert's top matches:
  - Emma with a score of 5.0
  - Sophia with a score of 5.0
Peter's top matches:
  - Isabella with a score of 5.0
  - Mary with a score of 4.0
Eric's top matches:
  - Martha with a score of 4.0
  - Sophia with a score of 4.0
Daniel's top matches:
  - Emma with a score of 8.0
  - Olivia with a score of 2.0
Adam's top matches:
  - Jennifer with a score of 6.0
  - Sophia with a score of 6.0
Carl's top matches:
  - Laura with a score of 8.0
  - Olivia with a score of 7.0
Ralph's top matches:
  - Isabella with a score of 8.0
  - Ol